### PR TITLE
[Data] Enable optimizer by default

### LINKED ISSUE #32596: [Datasets] Enabling execution optimizer after passing all CI and nightly tests
### Description

As title.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
This PR enables the execution plan optimizer in Ray Data, and fixes some bugs discovered via unit tests. We will ensure that Data CI and release tests are healthy before merging.

## Related issue number
Closes #32596

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/legacy_compat.py [hunk 1]
```diff
@@ -137,7 +137,11 @@ def _get_execution_dag(
         record_operators_usage(plan._logical_plan.dag)
 
     # Get DAG of physical operators and input statistics.
-    if DataContext.get_current().optimizer_enabled:
+    if (
+        DataContext.get_current().optimizer_enabled
+        and hasattr(plan, "_logical_plan")
+        and plan._logical_plan is not None
```
Comment (amogkam) [inline python/ray/data/_internal/execution/legacy_compat.py:143]: why is this change needed? add a comment about this?

File: python/ray/data/_internal/logical/rules/operator_fusion.py [hunk 2]
```diff
@@ -108,10 +115,13 @@ def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
 
         # We currently only support fusing for the following cases:
         # - MapOperator -> MapOperator
-        # - MapOperator -> AllToAllOperator
+        # - TaskPoolMapOperator -> AllToAllOperator
```
Comment (amogkam) [inline python/ray/data/_internal/logical/rules/operator_fusion.py:118]: why isn't actorpoolmapoperator -> alltoallloperator not supported? add a comment about this?

**Comment by raulchen** [comment]
@bveeramani Sorry, I just took over this PR from @scottjlee last Friday. I am trying to figure out the changes as well.

File: python/ray/data/context.py [hunk 3]
```diff
@@ -98,9 +98,8 @@
 DEFAULT_AUTO_LOG_STATS = False
 
 # Whether to enable optimizer.
-DEFAULT_OPTIMIZER_ENABLED = bool(
-    int(os.environ.get("RAY_DATA_NEW_EXECUTION_OPTIMIZER", "0"))
-)
+DEFAULT_OPTIMIZER_ENABLED = True
+# DEFAULT_OPTIMIZER_ENABLED = False
```
Comment (amogkam) [inline python/ray/data/context.py:102]: should we still keep the flag, but just set it to "1" by default?

File: python/ray/data/tests/test_all_to_all.py [hunk 4]
```diff
@@ -1771,7 +1773,7 @@ def get_node_id():
     node1_id = ray.get(get_node_id.options(resources={"bar:1": 1}).remote())
     node2_id = ray.get(get_node_id.options(resources={"bar:2": 1}).remote())
 
-    ds = ray.data.range(100, parallelism=2).random_shuffle()
+    ds = ray.data.range(100, parallelism=2).random_shuffle().materialize()
```
Comment (amogkam) [inline python/ray/data/tests/test_all_to_all.py:1776]: do these tests not work without adding `materialize`? this seems like a regression?

File: python/ray/data/tests/test_stats.py [hunk 5]
```diff
@@ -308,7 +316,7 @@ def test_dataset__repr__(ray_start_regular_shared):
     assert canonicalize(repr(ds2._plan.stats().to_summary())) == (
         "DatasetStatsSummary(\n"
         "   dataset_uuid=U,\n"
-        "   base_name=MapBatches(<lambda>),\n"
+        "   base_name=DoRead->MapBatches,\n"
```
Comment (ericl) [inline python/ray/data/tests/test_stats.py:319]: What are the reasons for these stats changes? Are we fusing things we didn't before the optimizer flip?

File: python/ray/data/_internal/logical/rules/operator_fusion.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:118]: Correct me if I'm wrong. For `map -> AllToAll`, we are going to fuse them as one AllToAll. But AllToAll doesn't support running with actors. Will add a comment.

Comment (amogkam) [reply python/ray/data/_internal/logical/rules/operator_fusion.py:118]: Got it, so this was a bug before

File: python/ray/data/tests/test_all_to_all.py (hunk 4 above)
Comment (ericl) [reply python/ray/data/tests/test_all_to_all.py:1776]: Yeah I agree, I would expect to see no test code changes for this PR.

File: python/ray/data/tests/test_stats.py (hunk 5 above)
Comment (raulchen) [reply python/ray/data/tests/test_stats.py:319]: At least for this case, the fusion behavior is the same. It looks that previously, for a fused operator, the name doesn't display the full chain, but just the last op name.

File: python/ray/data/tests/test_all_to_all.py (hunk 4 above)
Comment (raulchen) [reply python/ray/data/tests/test_all_to_all.py:1776]: I think they are just for debugging. Removed them. The tests still pass.

File: python/ray/data/context.py (hunk 3 above)
Comment (raulchen) [reply python/ray/data/context.py]: yeah, already did that.

File: python/ray/data/_internal/execution/legacy_compat.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/legacy_compat.py]: it's because some operators (union and window) doesn't support logical plan yet. Added a comment.

**Comment by raulchen** [comment]
@ericl @amogkam I've address all the comments, fixed all the known test failures left by Scott, and also reverted all the test changes that I think are not needed.

File: python/ray/data/__init__.py [hunk 6]
```diff
@@ -48,8 +48,11 @@
 
 # Module-level cached global functions for callable classes. It needs to be defined here
 # since it has to be process-global across cloudpickled funcs.
-_cached_fn = None
-_cached_cls = None
+# Used for caching user-defined callable classes.
+# Key the class, value is the object.
+# see make_callable_class_concurrent in python/ray/data/_internal/execution/util.py.
+# The reason why this is a dict is because we may fuse multiple map operators into one.
```
Comment (ericl) [inline python/ray/data/__init__.py:54]: Do we actually fuse multiple actors into one? I don't think we do that / should do that.

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 7]
```diff
@@ -31,13 +30,7 @@ def get_input_data() -> List[RefBundle]:
...
                         # tasks other than ray.put().
                         ray.put(read_task),
-                        BlockMetadata(
-                            num_rows=1,
-                            size_bytes=len(cloudpickle.dumps(read_task)),
-                            schema=None,
-                            input_files=[],
-                            exec_stats=None,
-                        ),
+                        read_task.get_metadata(),
```
Comment (ericl) [inline python/ray/data/_internal/planner/plan_read_op.py:33]: Can you make sure this calls `cleaned_metadata(read_task)` in legacy_compat.py to implement the same logic?

File: python/ray/data/tests/test_optimize.py [hunk 8]
```diff
@@ -308,10 +309,15 @@ def test_stage_linking(ray_start_regular_shared):
...
     _assert_has_stages(ds._plan._stages_before_snapshot, ["Map"])
     assert len(ds._plan._stages_after_snapshot) == 0
+    # since the new optimizer path does not use the stage fusion path,
+    # but instead performs operator fusion, we expect _last_optimized_stages
+    # to be not updated (None)?
     _assert_has_stages(ds._plan._last_optimized_stages, ["ReadRange->Map"])
 
 
 def test_optimize_reorder(ray_start_regular_shared):
+    # The ReorderRandomizeBlocksRule optimizer rule collapses RandomizeBlocks operators,
```
Comment (ericl) [inline python/ray/data/tests/test_optimize.py:319]: Shouldn't it still show up in the stage names?

**Comment by raulchen** [comment]
As this PR is getting huge, I'll split it into a few small PRs: the first two are #35621 and #35648.
I'll keep this PR open for testing purpose only. 

Other known issues not fixed in these 2 PRs include:
- Actors with the same class & args can be fused when not enabling optimizer, but not fused when enabling optimizer (need to identify necessity).
- `__repr__()` and `stats()` are different. 
- Some config flags (i.e., context.optimize_fuse_read_stages) are not respected.
